internal/core/items.rs

```diff
@@ -831,10 +843,42 @@ impl Item for Opacity {
     fn render(
         self: Pin<&Self>,
         backend: &mut ItemRendererRef,
-        _self_rc: &ItemRc,
+        self_rc: &ItemRc,
     ) -> RenderingResult {
-        backend.apply_opacity(self.opacity());
-        RenderingResult::ContinueRenderingChildren
+        backend.visit_opacity(self, self_rc)
+    }
+}
+
+impl Opacity {
+    // The opacity item has only one child, as per the compiler's output. However if that item has
+    // no children, then we can skip the layer and apply the opacity directly. This is not perfect though,
+    // for example if the compiler inserts another synthetic element between the `Opacity` and the actual child,
+    // then this check will apply a layer even though it might not actually be necessary.
+    pub fn opacity_needs_layer(self_rc: &ItemRc) -> bool {
+        let target_has_children = {
+            let component_rc = self_rc.component();
+            let component_ref = vtable::VRc::borrow_pin(&component_rc);
+            let self_index = self_rc.index();
+            // TODO: use first_child() once it exists
+            let item_tree = component_ref.as_ref().get_item_tree();
+            let target_item_index = match item_tree.as_slice()[self_index] {
+                crate::item_tree::ItemTreeNode::Item {
+                    children_count: _children_count,
+                    children_index,
+                    ..
+                } => {
+                    debug_assert_eq!(_children_count, 1);
+                    children_index as usize
+                }
+                crate::item_tree::ItemTreeNode::DynamicTree { .. } => panic!(
+                    "internal error: compiler placed Opacity element as a parent for a repeater!"
+                ),
+            };
+            let target_item = ItemRc::new(component_rc.clone(), target_item_index);
+            target_item.children_count() != Some(0)
+        };
+
+        target_has_children
```

Comment:
I guess the best here would be to visit the children of the children. and see if there is overlap.
Then again, we don't know what item really has contents or if they are just transparent TouchArea or other nodes.

Maybe we can leave this optimization out for now.

Reply:
Sure, I'll remove the commit. No objections there :)